File: libraries/ndf/ndf1Vmod.c
#include <stdlib.h>
#include "sae_par.h"
#include "ndf_err.h"
#include "ndf1.h"
#include "ndf_ast.h"
#include "mers.h"

void ndf1Vmod( const char *mode, char *vmode, size_t vmode_length, int *status ){
/*
*+
*  Name:
*     ndf1Vmod

*  Purpose:
*     Validate an access mode string.

*  Synopsis:
*     void ndf1Vmod( const char *mode, char *vmode, size_t vmode_length,
*                    int *status )

*  Description:
*     This function validates an access mode string, returning an upper
*     case version if it is valid. Otherwise, an error is reported.

*  Parameters:
*     mode
*        Pointer to a null terminated string holding the access mode string
*        to be validated. Valid values are "READ", "UPDATE" or "WRITE"
*        (case insensitive).
*     vmode
*        Pointer to an array in which to return a null terminated string
*        holding the validated access mode string in upper case (not
*        returned if the "mode" value is invalid).
*     vmode_length
*        The length of the supplied 'vmode' array. This should include
*        room for the terminating null.
*     *status
*        The global status.

*  Copyright:
*     Copyright (C) 2018 East Asian Observatory
*     All rights reserved.

*  Licence:
*     This program is free software; you can redistribute it and/or modify
*     it under the terms of the GNU General Public License as published by
*     the Free Software Foundation; either version 2 of the License, or (at
*     your option) any later version.
*
*     This program is distributed in the hope that it will be useful,but
*     WITHOUT ANY WARRANTY; without even the implied warranty of
*     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
*     General Public License for more details.
*
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 51 Franklin Street,Fifth Floor, Boston, MA
*     02110-1301, USA

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK)
*     DSB: David S. Berry (EAO)

*  History:
*     xxx (DSB):
*        Original version, based on equivalent Fortran function by RFWS.

*-
*/

/* Check inherited global status. */
   if( *status != SAI__OK ) return;

/* Check the access mode string supplied against each permitted value in
   turn, returning the appropriate validated value. */

/* ...READ access. */
   if( astChrMatch( mode, "READ" ) ) {
      ndf1Ccpy( "READ", vmode, vmode_length, status );

/* ...UPDATE access. */
   } else if( astChrMatch( mode, "UPDATE" ) ) {
      ndf1Ccpy( "UPDATE", vmode, vmode_length, status );

/* ...WRITE access. */
   } else if( astChrMatch( mode, "WRITE" ) ) {
      ndf1Ccpy( "WRITE", vmode, vmode_length, status );

/* If the access mode value was not recognised, then report an error. */
   } else {
      *status = NDF__MODIN;
      msgSetc( "BADMODE", mode );
      errRep( " ", "Invalid access mode '^BADMODE' specified (possible "
              "programming error).", status );
   }

/* Call error tracing function and exit. */
   if( *status != SAI__OK ) ndf1Trace( "ndf1Vmod", status );

}

File: libraries/ndf/ndf1_vmod.f
      SUBROUTINE NDF1_VMOD( MODE, VMODE, STATUS )
*+
*  Name:
*     NDF1_VMOD

*  Purpose:
*     Validate an access mode string.

*  Language:
*     Starlink Fortran 77

*  Invocation:
*     CALL NDF1_VMOD( MODE, VMODE, STATUS )

*  Description:
*     The routine validates an access mode string, returning an upper
*     case version if it is valid. Otherwise, an error is reported.

*  Arguments:
*     MODE = CHARACTER * ( * ) (Given)
*        The access mode string to be validated. Valid values are
*        'READ', 'UPDATE' or 'WRITE' (case insensitive).
*     VMODE = CHARACTER * ( * ) (Returned)
*        The validated access mode string in upper case (not returned if
*        the MODE value is invalid).
*     STATUS = INTEGER (Given and Returned)
*        The global status.

*  Algorithm:
*     -  Check the access mode value supplied against each permitted
*     value in turn, returning the appropriate validated value.
*     -  If the access mode was not recognised, then report an error.

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK)
*     {enter_new_authors_here}

*  History:
*     6-OCT-1989 (RFWS):
*        Original version.
*     {enter_changes_here}

*  Bugs:
*     {note_any_bugs_here}

*-
      
*  Type Definitions:
      IMPLICIT NONE              ! No implicit typing

*  Global Constants:
      INCLUDE 'SAE_PAR'          ! Standard SAE constants
      INCLUDE 'NDF_ERR'          ! NDF_ error codes

*  Arguments Given:
      CHARACTER * ( * ) MODE

*  Arguments Returned:
      CHARACTER * ( * ) VMODE

*  Status:
      INTEGER STATUS             ! Global status

*  External References:
      LOGICAL CHR_SIMLR          ! Case insensitive string comparison

*.

*  Check inherited global status.
      IF ( STATUS .NE. SAI__OK ) RETURN

*  Check the access mode string supplied against each permitted value in
*  turn, returning the appropriate validated value.

*  ...READ access.
      IF ( CHR_SIMLR( MODE, 'READ' ) ) THEN
         CALL NDF1_CCPY( 'READ', VMODE, STATUS )

*  ...UPDATE access.
      ELSE IF ( CHR_SIMLR( MODE, 'UPDATE' ) ) THEN
         CALL NDF1_CCPY( 'UPDATE', VMODE, STATUS )

*  ...WRITE access.
      ELSE IF ( CHR_SIMLR( MODE, 'WRITE' ) ) THEN
         CALL NDF1_CCPY( 'WRITE', VMODE, STATUS )

*  If the access mode value was not recognised, then report an error.
      ELSE
         STATUS = NDF__MODIN
         CALL MSG_SETC( 'BADMODE', MODE )
         CALL ERR_REP( 'NDF1_VMOD_BAD',
     :   'Invalid access mode ''^BADMODE'' specified (possible ' //
     :   'programming error).', STATUS )
      END IF
       
*  Call error tracing routine and exit.
      IF ( STATUS .NE. SAI__OK ) CALL NDF1_TRACE( 'NDF1_VMOD', STATUS )

      END
